### PR TITLE
Add "Getting Started" documentation for tokens

### PR DESCRIPTION
**Pull request by awrichar**
Initial docs for #218.

~~Some of the routes described here are still being updated by @eberger727 in #307, so should wait for that PR before merging this.~~

I've published the docs in this PR to my own FireFly fork here, if you like reading the pretty pages instead of the Markdown: https://awrichar.github.io/firefly/gettingstarted/mint_tokens.html

### CODE REVIEW

File: docs/gettingstarted/mint_tokens.md [hunk 1]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
 
 ---
 
-## Work in progress
+## Quick reference
 
-... the tokens APIs are under development
\ No newline at end of file
+- FireFly provides an abstraction layer for multiple types of tokens
+- Tokens are grouped into _pools_, which each represent a particular type or class of token
```
Comment (peterbroadhurst) [inline docs/gettingstarted/mint_tokens.md:22]: Wonder if an example worthwhile here, such as?

Examples of how a pool might be used by a given token connector implementation:
    - ERC-721: A contract instance, ready for issuance of NFTs within unique identities in the pool
    - ERC-20 / ERC-777: A token contract instance, with a single fungible pool of value - "a coin"
    - ERC-1155: An efficient to allocate isolated pool of fungible or non-fungible tokens (see [ref implementation](https://github.com/hyperledger/firefly-tokens-erc1155))
    - ERC-1400 (and related security token standards): An partition within an overall fungible pool, representing part of the value of an overall non-fungible asset

Could do with @jimthematrix checking on the terminology above

File: docs/gettingstarted/mint_tokens.md [hunk 2]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
+  "name": "testpool",
+  "type": "fungible"
+}
+```
+
+Other parameters:
+- You must specify a `connector` if you have configured multiple token connectors
+- You may pass through a `config` object of additional parameters, if supported by your token connector
+- You may specify a `key` understood by the connector (i.e. an Ethereum address) if you'd like to use a non-default signing identity
+
```
Comment (peterbroadhurst) [inline docs/gettingstarted/mint_tokens.md:48]: Worth noting that individual connector implementations an extend this without any updates to FireFly core, by using an opaque configuration setting. For example a token connector could allow an initial supply to be specified, or configuration options specifying transfer restriction criteria.

File: docs/gettingstarted/mint_tokens.md [hunk 3]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
+  "amount": 1,
+  "to": "0x07eab7731db665caf02bc92c286f51dea81f923f"
+}
+```
+
+Other parameters:
+- You must specify a `pool` name if you've created more than one pool
+- You must specify a `tokenIndex` for non-fungible pools (and the `amount` should be 1)
+- You may specify a `key` understood by the connector (i.e. an Ethereum address) if you'd like to use a non-default signing identity
+- You may specify `from` if you'd like to send tokens from a specific identity (default is the same as `key`)
```
Comment (peterbroadhurst) [inline docs/gettingstarted/mint_tokens.md:86]: @jimthematrix - note that `from` here is  the source account from which the funds should originate.
Whereas `key` is the signing key with which the transaction will be authorized by the on-chain contract.

I believe this is consistent throughout FireFly, but I wanted to call this out as the concrete is setting here, to check you agree this is acceptable and consistent.

**Review by peterbroadhurst — COMMENTED** [review]
Really clear and straight forwards thanks @awrichar 

Couple of comments, and deferral to @jimthematrix for a final review on this one before it goes in.

**Comment by awrichar** [comment]
Should I add a section on the FireFly configuration file? ie show the sample tokens config that is generated by the CLI?

File: docs/gettingstarted/mint_tokens.md (hunk 1 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md:22]: Could be, and maybe not limited to pools... Maybe we need an extra section of "Example uses" to lay out a few examples of how you could map common token standards/use cases onto FireFly.

File: docs/gettingstarted/mint_tokens.md (hunk 2 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md:48]: Yea, that is the `config` parameter mentioned above, but maybe we should give examples of that type of thing?

Comment (peterbroadhurst) [reply docs/gettingstarted/mint_tokens.md:48]: Sorry I missed that, looks good for now to me

File: docs/gettingstarted/mint_tokens.md (hunk 1 above)
Comment (jimthematrix) [reply docs/gettingstarted/mint_tokens.md:22]: if we agree with the above, we should also add that the current API endpoints support the ERC20, ERC777, ERC1155 and ERC721 contracts, while future work is needed to support ERC1400, in particular to support partitions (metadata to fungible tokens) within pools.

File: docs/gettingstarted/mint_tokens.md [hunk 4]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
-## Work in progress
+## Quick reference
 
-... the tokens APIs are under development
\ No newline at end of file
+- FireFly provides an abstraction layer for multiple types of tokens
+- Tokens are grouped into _pools_, which each represent a particular type or class of token
+- Each pool is classified as _fungible_ or _non-fungible_
+- In the case of _non-fungible_ tokens, the pool is subdivided into individual tokens with a unique _token index_
+- Within a pool, you may _mint_, _transfer_, and _burn_ tokens
```
Comment (jimthematrix) [inline docs/gettingstarted/mint_tokens.md:25]: suggest adding operational aliases (issue vs. mint, redeem vs. burn) used in different token specs:

Within a pool, you may _mint_ **(_issue_)**, _transfer_, and _burn_ **(_redeem_)** tokens

File: docs/gettingstarted/mint_tokens.md (hunk 3 above)
Comment (jimthematrix) [reply docs/gettingstarted/mint_tokens.md:86]: thanks, I think it's consistent with the lingo used by the various token specs (my earlier concern was that it's unfortunate that the transfer source account is designated as `from` which is also the parameter name of web3 interface's tx signing account, same for `to`. but I think this usage is part of the widely adopted convention now)

File: docs/gettingstarted/mint_tokens.md [hunk 5]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
+  "to": "0x07eab7731db665caf02bc92c286f51dea81f923f",
+  "message": {
+    "data": [{
+      "value": "payment for goods"
+    }]
+  }
+}
+```
+
+By default, a broadcast message is used. In order to send a private message, specify `"type": "transfer_private"` in the message header,
```
Comment (jimthematrix) [inline docs/gettingstarted/mint_tokens.md:112]: a sample payload for a private message would be useful here

File: docs/gettingstarted/mint_tokens.md [hunk 6]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
+
+Other parameters:
+- You must specify a `pool` name if you've created more than one pool
+- You must specify a `tokenIndex` for non-fungible pools (and the `amount` should be 1)
+- You may specify a `key` understood by the connector (i.e. an Ethereum address) if you'd like to use a non-default signing identity
+- You may specify `from` if you'd like to send tokens from a specific identity (default is the same as `key`)
+
+## Sending data with a transfer
+
+All transfers (as well as mint/burn operations) support an optional `message` parameter that contains a broadcast or private
```
Comment (jimthematrix) [inline docs/gettingstarted/mint_tokens.md:90]: I think we effectively hijacked the `_data` field of the various operations that may have been needed for further on-chain authorizations (e.g signatures by parties other than the sender to indicate approval). as a mental note we should consider supporting an additional payload parameter to include into the `_data` field as-is

**Review by jimthematrix — CHANGES_REQUESTED** [review]
Nice writeup. Added my opinion on support for ERC-1400, and other relatively minor comments. Happy to discuss more if there are different thoughts.

File: docs/gettingstarted/mint_tokens.md [hunk 7]
```diff
@@ -16,6 +16,118 @@ nav_order: 9
...
+- You must specify a `connector` if you have configured multiple token connectors
+- You may pass through a `config` object of additional parameters, if supported by your token connector
+- You may specify a `key` understood by the connector (i.e. an Ethereum address) if you'd like to use a non-default signing identity
+
+## Mint tokens
+
+Once you have a token pool, you can mint tokens within it. With the default `firefly-tokens-erc1155` connector,
+only the creator of a pool is allowed to mint - but each connector may define its own permission model.
+
+`POST` `/api/v1/namespaces/default/tokens/mint`
```
Comment (nguyer) [inline docs/gettingstarted/mint_tokens.md:54]: I think the token pool name part is missing from this path

Comment (nguyer) [reply docs/gettingstarted/mint_tokens.md:54]: Okay actually I see now that it just works if there's only one pool defined. Does this URL break if there is more than one pool?

**Review by nguyer — APPROVED** [review]
Looks great. I ran through the instructions in Postman and it worked for me too. Jim has requested changes, but if he dismisses or approves after updates, I'm fine with this being merged. Thanks!

File: docs/gettingstarted/mint_tokens.md (hunk 7 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md:54]: The pool name is not in the new paths (if there is only one pool the name is optional; if there is more than one pool, you pass it in the body). The old path with the pool name is deprecated.

File: docs/gettingstarted/mint_tokens.md (hunk 6 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md:90]: @jimthematrix yes, you're correct. The token connector exposes a `data` field, and FireFly packs JSON into this field to identify messages and other metadata. But FireFly itself does not (currently) expose any way for the user to control or add to this field from the level of the FireFly APIs.

File: docs/gettingstarted/mint_tokens.md (hunk 1 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md:22]: Added a new section with some of these examples.

File: docs/gettingstarted/mint_tokens.md (hunk 5 above)
Comment (awrichar) [reply docs/gettingstarted/mint_tokens.md]: added

**Review by jimthematrix — APPROVED** [review]

**Pull request merged**